bmcd/src/legacy.rs

```diff
@@ -0,0 +1,395 @@
+}
+
+fn set_node_info() -> HttpResponse {
+    // In previous versions of the firmware this was dead code
+    HttpResponse::NotImplemented().body("Method type `set` on parameter `nodeinfo` is deprecated")
+}
+
+fn get_node_info(_bmc: &mut Arc<BmcApplication>) -> HttpResponse {
+    // TODO: implement serial listening in BmcApplication
+    let (n1, n2, n3, n4) = (0, 0, 0, 0);
+
+    let body = json!({
+        "response": [{
+            "node1": n1,
+            "node2": n2,
+            "node3": n3,
+            "node4": n4,
+        }]
+    });
+
+    HttpResponse::Ok().json(body)
+}
+
+async fn set_node_to_msd(bmc: &mut Arc<BmcApplication>, query: Query) -> HttpResponse {
+    let node = match get_node_param(&query) {
+        Ok(n) => n,
+        Err(e) => return HttpResponse::BadRequest().body(e),
+    };
+
+    let (tx, mut rx) = mpsc::channel(64);
+
+    let bmc = bmc.clone();
+    let handle = rt::spawn(async move { bmc.set_node_in_msd(node, UsbRoute::Bmc, tx).await });
+
+    let logger = rt::spawn(async move {
+        while let Some(msg) = rx.recv().await {
+            log::info!("{}", msg);
+        }
+    });
+
+    tokio::try_join!(handle, logger).to_response("await threads")
+}
+
+fn get_node_param(query: &Query) -> Result<NodeId, &'static str> {
+    let Some(node_str) = query.get("node") else {
+        return Err("Missing `node` parameter");
+    };
+
+    let Ok(node_num) = i32::from_str(node_str) else {
+        return Err("Parameter `node` is not a number");
+    };
+
+    let Ok(node) = node_num.try_into() else {
+        return Err("Parameter `node` is out of range 0..3 of node IDs");
+    };
+
+    Ok(node)
+}
+
+async fn get_system_information() -> HttpResponse {
+    let version = env!("CARGO_PKG_VERSION");
+    let build_time = build_time::build_time_utc!("%Y-%m-%d %H:%M:%S-00:00");
+    let ipv4 = get_ipv4_address().unwrap_or("Unknown".to_owned());
+    let mac = get_mac_address().await;
+
+    let body = json!({
+        "response": [{
+            "version": version,
+            "buildtime": build_time,
+            "ip": ipv4,
+            "mac": mac,
+        }]
+    });
+
+    HttpResponse::Ok().json(body)
+}
+
+fn get_ipv4_address() -> Option<String> {
+    for interface in if_addrs::get_if_addrs().ok()? {
+        // TODO: for compatibility reasons, only IPv4 of eth0 is returned. Ideally, both IPv4 and
+        // IPv6 addresses of all non-loopback interfaces should be returned.
+        if interface.is_loopback() || interface.name != "eth0" {
+            continue;
+        }
+
+        let std::net::IpAddr::V4(ip) = interface.ip() else {
+            continue;
+        };
+
+        return Some(format!("{}", ip));
+    }
+
+    None
+}
+
+async fn get_mac_address() -> String {
+    tokio::fs::read_to_string("/sys/class/net/eth0/address")
+        .await
+        .unwrap_or("Unknown".to_owned())
+}
+
+async fn set_node_power(bmc: &mut Arc<BmcApplication>, query: Query) -> HttpResponse {
```

Comment:
I want to propose to change behavior here a bit, as i think the old code was not necessarily great here:
currently, the function is setup to return a `BadRequest()` when one of the `nodeX=1` parameters is missing. It would be nice if this function could walk over the parameters in the `Query` and gather all nodes to update, as a last step it will call `bmc.activate_slot` at once.

For example, i expect the following query to be handled:
```node3=1&node=0```
which would result in bmc call:

```rust
bmc.activate_slot(1000u8,1001u8).await
```

we do internal mutability, try to keep the `BmcApplication` interface immutable

 ```suggestion
async fn set_node_power(bmc: &Arc<BmcApplication>, query: Query) -> HttpResponse {
```

Reply:
It does actually allow some parameters to be missing, it returns `BadRequest` when the value of a given paramter is malformed, e.g. `node1=1&node2=asdf`.

Reply:
I read over the `continue` so you are right. However, I would still prefer only one call to `activate_slot`. I suggest defining a node and mask variable outside the loop and update its values while you are parsing the query. As a last step, you can call `activate_slot` with the final values